function-server/main.py

```diff
@@ -47,10 +52,19 @@ def process_msg(msg, handle):
     try:
         sys.stderr = stderr
         sys.stdout = stdout
-        r = handle(msg['context'], msg['payload'])
+        with ThreadPoolExecutor(max_workers=1) as executor:
+            future = executor.submit(handle, msg['context'], msg['payload'])
+            timeout = msg['context'].get('timeout', None)
+            r = future.result(timeout / 1000)
     except (ValueError, TypeError) as e:
         stacktrace = traceback.format_exc().splitlines()
         err = {'type': INPUT_ERROR, 'message': str(e), 'stacktrace': stacktrace}
+    except TimeoutError as e:
+        stacktrace = traceback.format_exc().splitlines()
+        err = {'type': SYSTEM_ERROR, 'message': str(e), 'stacktrace': stacktrace}
+        global HEALTHY
```

Comment:
If we just cancel the future, why do we need to mark this executor as unhealthy?

Reply:
There is no guarantee that the future can be cancelled. The function itself would need to know how to respond to the cancel, which would allow the function to properly release resources. I've updated it so that it only sets unhealthy in the cancel fails.